File: src/0383-Ransom-Note/ransom_note.py
from typing import Dict

class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        char_map: Dict = {}

        for char in magazine:
            char_map[char] = char_map.get(char, 0) + 1

        for char in ransomNote:
            if char not in char_map or char_map[char] == 0:
                return False
            char_map[char] -= 1

        return True
 
if __name__ == "__main__":
    magazine: str = "ab"
    ransomNote: str = "aa"
    solution = Solution()

    print(solution.canConstruct(ransomNote, magazine))

File: src/0383-Ransom-Note/ransom_note.cpp
#include <iostream>
#include <unordered_map>

class Solution {
public:
    bool canConstruct(const std::string& ransomNote, const std::string& magazine) {
        std::unordered_map<char, int> char_map;

        for (char c : magazine) {
            char_map[c] = char_map[c] + 1;
        }

        for (char c : ransomNote) {
            if (char_map.find(c) == char_map.end() || char_map[c] == 0) {
                return false;
            }
            char_map[c] = char_map[c] - 1;
        }

        return true;
    }
};

int main() {
    std::string magazine = "ab";
    std::string ransomNote = "aa";
    Solution solution;

    std::cout << std::boolalpha << solution.canConstruct(ransomNote, magazine) << std::endl;

    return 0;
}